mirrord/layer/src/file/hooks.rs

```diff
@@ -985,6 +985,73 @@ unsafe extern "C" fn realpath_darwin_extsn_detour(
 //     result
 // }
 
+unsafe fn vec_to_iovec(bytes: &Vec<u8>, iovecs: &[iovec]) {
+    let mut copied = 0;
+    let mut iov_index = 0;
+
+    while copied < bytes.len() {
+        let iov = &iovecs.get(iov_index).expect("ioevec out of bounds");
+        let read_ptr = bytes.as_ptr();
+        let copy_amount = std::cmp::min(bytes.len(), iov.iov_len);
+        let out_buffer = iov.iov_base.cast();
+        ptr::copy(read_ptr, out_buffer, copy_amount);
+        copied += copy_amount;
+        // we trust iov_index to be in correct size since we checked it before
+        iov_index += 1;
+    }
+}
+/// Hook for `libc::readv`.
+#[hook_guard_fn]
+pub(crate) unsafe extern "C" fn readv_detour(
+    fd: RawFd,
+    iovecs: *const iovec,
+    iovec_count: c_int,
+) -> ssize_t {
+    if iovec_count < 0 {
+        return FN_READV(fd, iovecs, iovec_count);
+    }
+    let iovs = slice::from_raw_parts(iovecs, iovec_count as usize);
```

Comment:
Will probably segfault if `iovecs` is `null` (well, not here, but actually when we try to do stuff with this pointer)?

What is normal `libc` behavior when this happens?

Reply:
I don't think it does null checks

Reply:
```
readv: Bad address
```

Original `readv` returns this when I call `readv(fd, NULL, 3)`.